src/lib.rs

```diff
@@ -301,6 +301,78 @@ where
         self.inner.remove(k)
     }
 
+    /// Removes the first item in the vector corresponding to the key from the map.
+    ///
+    /// The key may be any borrowed form of the map's key type, but Hash and Eq
+    /// on the borrowed form must match those for the key type.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use multimap::MultiMap;
+    ///
+    /// let mut map = MultiMap::new();
+    /// map.insert(1, 42);
+    /// map.insert(1, 1337);
+    /// assert_eq!(map.remove_first(&1), Some(42));
+    /// assert_eq!(map.remove_first(&1), Some(1337));
+    /// assert_eq!(map.remove_first(&1), None);
+    /// ```
+    pub fn remove_first<Q: ?Sized>(&mut self, k: &Q) -> Option<V>
+    where
+        K: Borrow<Q>,
+        Q: Eq + Hash,
+    {
+        let vec = self.get_vec_mut(k)?;
+        if vec.is_empty() {
+            // We shouldn't ever reach this branch but handle it for sanity
+            self.remove(k);
+            None
+        } else {
+            let item = vec.remove(0);
+
+            // If the entry vec is completely empty we can remove it
+            if vec.is_empty() {
```

Comment:
I don't really agree that this method should remove the empty underlying vec.

Reply:
Won't it leak memory otherwise? The empty entries will remain there even after the last item was removed.

Reply:
You're not wrong, it's just that I think this is sort of the expected behavior by rust collections. See [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6b7849795bcaca94be9520187964df1f) as an example of how ```Vec``` also "leaks" memory.

But, then again this isn't exactly the same either as this is one more "level".

Reply:
Also, it seems likely that the ```key``` will be reused and then you have to allocate new memory on the heap.